From fb62b057a317a46bb6c13fac402c2479c8687bbe Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Kacper=20Michaj=C5=82ow?= <kasper93@gmail.com>
Date: Wed, 23 Sep 2026 16:16:47 +0200
Subject: [PATCH] csputils: share the AVStereo3D to mp_stereo3d_mode
 mapping

Move it out of mp_image_from_av_frame() next to mp_stereo3d_names[],
so demuxers can map stream side data the same way.

(cherry picked from commit 1d36c85a9af23e5469705ee8a6b7e70e18c334cc)
---
 video/csputils.c | 24 ++++++++++++++++++++++++
 video/csputils.h |  3 +++
 video/mp_image.c | 25 +------------------------
 3 files changed, 28 insertions(+), 24 deletions(-)

diff --git a/video/csputils.c b/video/csputils.c
index d59e672708..4c9b63b729 100644
--- a/video/csputils.c
+++ b/video/csputils.c
@@ -23,6 +23,7 @@
 #include <math.h>
 #include <assert.h>
 #include <libavutil/common.h>
+#include <libavutil/stereo3d.h>
 #include <libavcodec/avcodec.h>
 
 #include "mp_image.h"
@@ -154,6 +155,29 @@ const struct m_opt_choice_alternatives mp_stereo3d_names[] = {
     {0}
 };
 
+enum mp_stereo3d_mode mp_stereo3d_from_av(const struct AVStereo3D *s3d)
+{
+    bool inv = s3d->flags & AV_STEREO3D_FLAG_INVERT;
+    switch (s3d->type) {
+    case AV_STEREO3D_SIDEBYSIDE:
+    case AV_STEREO3D_SIDEBYSIDE_QUINCUNX:
+        return inv ? MP_STEREO3D_SBS2R : MP_STEREO3D_SBS2L;
+    case AV_STEREO3D_TOPBOTTOM:
+        return inv ? MP_STEREO3D_AB2R : MP_STEREO3D_AB2L;
+    case AV_STEREO3D_CHECKERBOARD:
+        return inv ? MP_STEREO3D_CHECKR : MP_STEREO3D_CHECKL;
+    case AV_STEREO3D_LINES:
+        return inv ? MP_STEREO3D_IRR : MP_STEREO3D_IRL;
+    case AV_STEREO3D_COLUMNS:
+        return inv ? MP_STEREO3D_ICR : MP_STEREO3D_ICL;
+    case AV_STEREO3D_FRAMESEQUENCE:
+        return inv ? MP_STEREO3D_AR : MP_STEREO3D_AL;
+    default:
+        // AV_STEREO3D_2D and types mpv has no mode for.
+        return MP_STEREO3D_MONO;
+    }
+}
+
 void mp_get_3d_side_by_side(int stereo_mode, int div[2])
 {
     div[0] = div[1] = 1;
diff --git a/video/csputils.h b/video/csputils.h
index d53345114a..83e2618937 100644
--- a/video/csputils.h
+++ b/video/csputils.h
@@ -69,6 +69,9 @@ enum mp_stereo3d_mode {
 
 extern const struct m_opt_choice_alternatives mp_stereo3d_names[];
 
+struct AVStereo3D;
+enum mp_stereo3d_mode mp_stereo3d_from_av(const struct AVStereo3D *s3d);
+
 #define MP_STEREO3D_NAME_DEF(x, def) m_opt_choice_str_def(mp_stereo3d_names, x, def)
 
 // Number of screen divisions per axis (div[0]=x, div[1]=y) for the given 3D mode
diff --git a/video/mp_image.c b/video/mp_image.c
index d6f780554c..1f5e336aee 100644
--- a/video/mp_image.c
+++ b/video/mp_image.c
@@ -1079,29 +1079,6 @@ void mp_image_params_guess_csp(struct mp_image_params *params)
     }
 }
 
-static enum mp_stereo3d_mode stereo3d_from_av(const AVStereo3D *s3d)
-{
-    bool inv = s3d->flags & AV_STEREO3D_FLAG_INVERT;
-    switch (s3d->type) {
-    case AV_STEREO3D_SIDEBYSIDE:
-    case AV_STEREO3D_SIDEBYSIDE_QUINCUNX:
-        return inv ? MP_STEREO3D_SBS2R : MP_STEREO3D_SBS2L;
-    case AV_STEREO3D_TOPBOTTOM:
-        return inv ? MP_STEREO3D_AB2R : MP_STEREO3D_AB2L;
-    case AV_STEREO3D_CHECKERBOARD:
-        return inv ? MP_STEREO3D_CHECKR : MP_STEREO3D_CHECKL;
-    case AV_STEREO3D_LINES:
-        return inv ? MP_STEREO3D_IRR : MP_STEREO3D_IRL;
-    case AV_STEREO3D_COLUMNS:
-        return inv ? MP_STEREO3D_ICR : MP_STEREO3D_ICL;
-    case AV_STEREO3D_FRAMESEQUENCE:
-        return inv ? MP_STEREO3D_AR : MP_STEREO3D_AL;
-    default:
-        // AV_STEREO3D_2D and types mpv has no mode for.
-        return MP_STEREO3D_MONO;
-    }
-}
-
 // Create a new mp_image reference to av_frame.
 struct mp_image *mp_image_from_av_frame(struct AVFrame *src)
 {
@@ -1156,7 +1133,7 @@ struct mp_image *mp_image_from_av_frame(struct AVFrame *src)
 
     sd = av_frame_get_side_data(src, AV_FRAME_DATA_STEREO3D);
     if (sd)
-        dst->params.stereo3d = stereo3d_from_av((const AVStereo3D *)sd->data);
+        dst->params.stereo3d = mp_stereo3d_from_av((const AVStereo3D *)sd->data);
 
     if (src->opaque_ref) {
         struct mp_image_params *p = (void *)src->opaque_ref->data;
